File: plugin-rust-agent/src/main/java/jetbrains/buildServer/rust/logging/CargoStateLogger.java
/*
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * See LICENSE in the project root for license information.
 */

package jetbrains.buildServer.rust.logging;

import jetbrains.buildServer.agent.BuildProgressLogger;
import org.jetbrains.annotations.NotNull;

/**
 * State-based logger.
 */
public class CargoStateLogger extends CargoDefaultLogger {
    private static final String MESSAGE_FORMAT = "##teamcity[message text='%s']";
    private final BuildProgressLogger myLogger;
    private final CargoState myState;

    public CargoStateLogger(@NotNull final BuildProgressLogger logger, CargoState state) {
        super(logger);
        myLogger = logger;
        myState = state;
    }

    @Override
    public void onEnter(String text) {
        final String message = myState + " " + text;
        myLogger.message(String.format(MESSAGE_FORMAT, message));
    }

    @Override
    public void processLine(String text) {
        myLogger.message(String.format(MESSAGE_FORMAT, text));
    }

    @Override
    public void onLeave() {
    }
}

File: plugin-rust-agent/src/main/kotlin/jetbrains/buildServer/rust/logging/CargoStateLogger.kt


package jetbrains.buildServer.rust.logging

import jetbrains.buildServer.agent.BuildProgressLogger

/**
 * State-based logger.
 */
class CargoStateLogger(private val myLogger: BuildProgressLogger,
                       private val myState: CargoState,
                       private val myPriority: String? = null)
    : CargoDefaultLogger(myLogger) {

    private var myMessage: StringBuilder? = null

    override fun onEnter(text: String) {
        if (myPriority.isNullOrEmpty()) {
            myLogger.message(String.format(MESSAGE_FORMAT, "$myState ${escapeValue(text)}", ""))
        } else {
            myMessage = StringBuilder("$myState $text")
        }
    }

    override fun processLine(text: String) {
        if (myMessage == null) {
            myLogger.message(String.format(MESSAGE_FORMAT, escapeValue(text), ""))
        } else {
            myMessage?.append(text)?.append("\n")
        }
    }

    override fun onLeave() {
        if (myMessage != null) {
            val message = escapeValue(myMessage.toString().trim())
            myLogger.message(String.format(MESSAGE_FORMAT, message, " status='$myPriority'"))
            myMessage = null
        }
    }

    companion object {
        private const val MESSAGE_FORMAT = "##teamcity[message text='%s'%s]"
    }
}